W-14927671  Pdk ga branch gr (#464)
mulesoft/docs-gateway: pdk/1.0/modules/ROOT/pages/policies-pdk-configure-features-caching.adoc

> = Sharing Data Between Workers and Caching

Reviewer: ```suggestion
= Sharing Data Between Workers and the Caching Mechanism
```
Or something. Not sure. At first read I thought that this tutorial included steps for workers to share data amongst themselves, **and** steps for some kind of caching. The wording is a little ambiguous and therefore it confused me.

Or maybe:
"Sharing Worker Data with the Cache"?

Or maybe:
"Caching Worker Data"?

Again, not sure. Something to think about.

Author: https://github.com/mulesoft/docs-gateway/pull/504

ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]
:imagesdir: ../assets/images

NOTE: To view an example policy project that uses caching, see https://github.com/mulesoft/pdk-custom-policy-examples/tree/main/data-caching/README.md[Data Caching Policy Example^].

> Envoy spawns many workers for the policy to increase the capability of handling requests. A single worker handles all the stages of a specific request, including outgoing connections and the response flow, enabling you to track an internal status for a specific request and create a sequential execution flow even though the underlying implementation is event-driven. 

Reviewer: ```suggestion
Envoy spawns many workers to increase the policy's capability of handling requests. A single worker handles all the stages of a specific request, including outgoing connections and the response flow. This enables you to track internal status for a specific request, and to create a sequential execution flow despite the underlying event-driven implementation. 
```
Just some suggestions for wordiness and sentence simplification (and an instance of parallelization using "to".)

> Each worker processes a single request at a given time. To share a global state in a specific worker, use a `RefCell` without any risk of concurrent modification.

Reviewer: As a newbie, I'm not sure what a `RefCell` is. Is it a parameter in some file? Would it be self-explanatory to a new user? I don't see an example of it in the below snippets. Or can you link to an explanation if explained on another page?

Author: Rust data structure

> Use the provided `Cache` mechanism to communicate a shared global state to all workers.

Reviewer: ```suggestion
Use the provided `Cache` mechanism to create a shared global state for all workers.
```
I'm not sure the mechanism "communicates" a state? In my mind it "creates" a state, but I could be wrong.

Or maybe "establishes"? Not sure.

> After injecting the cache builder, provide an ID for the cache and the maximum number of elements the cache can support. 

Reviewer: What is "injecting the cache builder"? Is that the same as using the "provided `Cache` mechanism"? As a newbie, I'm not sure if I missed an "injecting" step, and the imperative sentence implies that I should have already done this.

Author: injecting is a fancy rust word for pass parameters, kind of standard

> In the following example, a cache with ID “calculator” can hold ten elements:

Reviewer: I don't see any mention of `calculator` in the following snippet. Is that supposed to be set somewhere?

Author: https://github.com/mulesoft/docs-gateway/pull/504/

[source,Rust]
----
#[entrypoint]
async fn configure(
    launcher: Launcher,
    cache_builder: CacheBuilder,
) -> Result<()> {
    let cache = cache_builder.new(String::from("caching")).max_entries(10);

>     ...

Reviewer: What else would potentially go here? As a newbie I'm not sure what I do with the ellipsis. Is this where my own custom configuration code will go? Is this explained in a different tutorial?

Maybe include a comment that says what will potentially go there?

I assume the code won't compile as is. Does it need to? If not, should we say that and explain that other code will need to be included?

Also, does the `Result` block need a closing `}`?

----

The cache has the following interface:

[source,Rust]
----
pub trait Cache {
    fn save(&self, key: &str, value: Vec<u8>) -> Result<(), CacheError>;
    fn get(&self, key: &str) -> Option<Vec<u8>>;
    fn delete(&self, key: &str) -> Option<Vec<u8>>;
    fn purge(&self);
}
----

> IMPORTANT: Because different works write to the cache concurrently, there is no guarantee that the value overwritten when saving the cache is the same as when it was retrieved.

Reviewer: ```suggestion
IMPORTANT: Because different workers write to the cache concurrently, there is no guarantee that the value overwritten when saving the cache is the same as when it was retrieved.
```
